File: Algorithms/44. Wildcard Matching.java
class Solution {
    private String s,p;
    public boolean isMatch(String s, String p) {
        this.s=s;
        this.p=p;
        boolean[][] f=new boolean[s.length()+1][p.length()+1];
        f[0][0]=true;
        for(int i=0;i<=s.length();i++){
            for(int j=1;j<=p.length();j++){
                if(p.charAt(j-1)=='*'){
                    f[i][j]=f[i][j-1] || ((i>0)&&(f[i-1][j] || f[i-1][j-1]));
                }
                else{
                    f[i][j]=(i>0) && f[i-1][j-1] && ((p.charAt(j-1)=='?')||(p.charAt(j-1)==s.charAt(i-1)));
                }
            }
        }
        return f[s.length()][p.length()];
    }
}

File: Algorithms/44. Wildcard Matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        dp = [[False for i in range(len(p)+1)] for i in range(len(s)+1)]
        dp[0][0] = True
        for i in range(0,len(s)+1):
            for j in range(1,len(p)+1):
                if p[j-1]=='*':
                    dp[i][j] = dp[i][j] or ((i>0) and dp[i-1][j]) or dp[i][j-1] or ((i>0) and dp[i-1][j-1])
                if i>0:
                    if s[i-1]==p[j-1] or p[j-1]=='?':
                        dp[i][j] = dp[i][j] or dp[i-1][j-1]
        return dp[len(s)][len(p)]